### DISCUSSION TITLE
Add per hit object information to output of difficultycalculator for display purposes only

### DISCUSSION THREAD

**Discussion by joseph-ireland** [ppy/osu#14934] [Ideas]
Difficulty calculations should be much more visible. Right now each map spits out a small handful of values, and it's hard to see where the results come from, i.e. which hit objects add value to the map and why.

One thing that could help immensely would be to be able to view the individual values for each hit object that is processed. It would be very instructive to see e.g. the angle bonus, aim and speed strain values, etc.

This should directly supported in the difficulty calculation framework - maybe even integrated into the editor in future.

My thoughts are that the DifficultyCalculator should have a separate function to do a "verbose" difficulty calculation, which would also return a list of intermediate values.

Writing the code for this seems a bit tricky due to the current structure, it would likely be quite a big PR to transform all of the existing code, but I think it would be worth it. In terms of implementation, I think it makes sense to store this stuff inside DifficultyHitObject, and add an option to return this list from DifficultyCalculator. 

To populate values I see two options:

1) Skills return a polymorphic SkillHitObjectAttributes containing intermediate values, they are stored in an array in DiffiultyHitObject

2) Skills directly fill values into the DifficultyHitObject

Number 2 seems easier to implement at first glance, but the way StrainSkill is abstracted would need to be refactored to allow access to the totalStrain inside each Skill - it probably makes sense to reverse the control, adding `Process` to each skill and an `AddStrain(double time, double strainValue)` to update the strain peaks, etc

**Comment by bdach** [comment]
I think an important consideration here is going to be making sure that none of the 'verbosity' code actually runs in production scenarios. There's a whole slew of reasons for this - memory considerations, the fact that they would not be used for anything useful, etc.

So I think something that may work better is giving all skills the access to a "logger" of some sort that they can use to write some value associated with a given difficulty hitobject. This logger can be substituted to a no-op in production scenarios but still yield the verbosity you want it to in testing.
